### PR TITLE
General refactoring

### PR DESCRIPTION
**Pull request by shegeley**
Some general refactoring. Was able to successfully pass `configure` phase via `guix package -f guix.scm`. The `build` phase is failing for now. May use help.

### CODE REVIEW

File: lisp/wayland/Makefile.am [hunk 1]
```diff
@@ -16,21 +17,33 @@ GUILE_WARNINGS =                                          \
   -Wformat
 
 SUFFIXES = .scm .go
+
 .scm.go:
-	$(AM_V_GEN)$(top_builddir)/pre-inst-env $(GUILE_TOOLS) compile $(GUILE_TARGET) $(GUILE_WARNINGS) -o "$@" "$<"
+	$(AM_V_GEN)$(top_builddir)/pre-inst-env $(GUILE_TOOLS) compile $(GUILE_TARGET)
+
+$(GUILE_WARNINGS) -o "$@" "$<"
```
Comment (Z572) [inline lisp/wayland/Makefile.am:25]: makefile's rule looks likes 
``` makefile
target: dependencies
   command
   ...
```

see https://web.mit.edu/gnu/doc/html/make_4.html

so 

``` makefile
.scm.go:
	$(AM_V_GEN)$(top_builddir)/pre-inst-env $(GUILE_TOOLS) compile $(GUILE_TARGET) \
        $(GUILE_WARNINGS) -o "$@" "$<"
```

**Comment by Z572** [comment]
Thans for you help, i will use some commit

File: Makefile.am [hunk 2]
```diff
@@ -27,6 +27,8 @@ TESTS= tests/display.scm
 
 AM_TESTS_ENVIRONMENT = abs_top_srcdir="$(abs_top_srcdir)"
 
+README: README.org
+
```
Comment (Z572) [inline Makefile.am:34]: no need add it for makefile, just use `ln -s README.org README` ,likes b165499

File: AUTHORS [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Original author: Z572@github.com (did 99.9% of work)
+Humble contributor: Grigory Shepelev | shegeley@github.com | (did the rest)
```
Comment (Z572) [inline AUTHORS:2]: email address should a real address, like me is 873216071@qq.com, and user can mailto me

File: repl.scm [hunk 4]
```diff
@@ -0,0 +1,2 @@
+;; TODO here add a script that will allow to launch the process locally with fully-functional repl (geiser) or whatever
+;; Starting a repl in is very simple ~guile ... -l <port>~
```
Comment (Z572) [inline repl.scm:2]: just `./pre-inst-env guile --listen`, and emacs use `geiser-connect`, and repl works

File: Makefile.am (hunk 2 above)
Comment (shegeley) [reply Makefile.am:34]: Hm... I hadn't have local (lisp/wayland/README) symlinked with the "top level one" and build system would throw an error. Maybe git doesn't track symlinks by default?

File: lisp/wayland/Makefile.am (hunk 1 above)
Comment (shegeley) [reply lisp/wayland/Makefile.am:25]: Thanks for really fast replies. I need guile wayland bindings (not all of them just a little part) for my personal project and it's the only suitable repo I found that still requires some DIY. 

I am very thankful for all the work you've done. All the formatting is totally up to my taste, I'm not judging the formatting of yours in any way. And I'm just started making the changed I want.  

Besides It's my first time into Make and Autotools.

Comment (shegeley) [reply lisp/wayland/Makefile.am:25]: I don't even insist on the PR being merged. Will convert to draft.

File: Makefile.am (hunk 2 above)
Comment (Z572) [reply Makefile.am:34]: The compilation system does not depend on git, `README: README.org` is tell Make `README` depend by `README.org`, but you no write how to create `README`, so Make cann't create `README`.

File: lisp/wayland/Makefile.am (hunk 1 above)
Comment (Z572) [reply lisp/wayland/Makefile.am:25]: > Thanks for really fast replies. I need guile wayland bindings (not all of them just a little part) for my personal project and it's the only suitable repo I found that still requires some DIY. 

if you need any miss wayland ffi binding, you can open a issue/PR, i will add it. 

> All the formatting is totally up to my taste, I'm not judging the formatting of yours in any way. And I'm just started making the changed I want.

This is not formatting issues, it's makefile's syntax error, https://github.com/guile-wayland/guile-wayland/blob/1ec967d427e67310135248d9a916735d595230ae/lisp/wayland/Makefile.am#L22-L23 fix it, and make `build` phase work.

File: repl.scm (hunk 4 above)
Comment (shegeley) [reply repl.scm:2]: won't work sadly. 
```
shegeley@zone51 ~/Projects/guile-wayland$ bash ./pre-inst-env.in -- guile 
./pre-inst-env.in: line 3: cd: @abs_top_srcdir@: No such file or directory
./pre-inst-env.in: line 4: cd: @abs_top_builddir@: No such file or directory
GNU Guile 3.0.9
Copyright (C) 1995-2023 Free Software Foundation, Inc.

Guile comes with ABSOLUTELY NO WARRANTY; for details type `,show w'.
This program is free software, and you are welcome to redistribute it
under certain conditions; type `,show c' for details.

Enter `,help' for help.
scheme@(guile-user)> ,m (wayland)
;;; note: auto-compilation is enabled, set GUILE_AUTO_COMPILE=0
;;;       or pass the --no-auto-compile argument to disable.
;;; compiling /home/shegeley/Projects/guile-wayland/lisp/wayland.scm
;;; compiling /home/shegeley/Projects/guile-wayland/lisp/wayland/client.scm
;;; compiling /home/shegeley/Projects/guile-wayland/lisp/wayland/server/client.scm
;;; compiling /home/shegeley/Projects/guile-wayland/lisp/wayland/display.scm
;;; WARNING: compilation of /home/shegeley/Projects/guile-wayland/lisp/wayland/display.scm failed:
;;; no code for module (bytestructure-class)
;;; WARNING: compilation of /home/shegeley/Projects/guile-wayland/lisp/wayland/server/client.scm failed:
;;; no code for module (wayland config)
;;; compiling /home/shegeley/Projects/guile-wayland/lisp/wayland/listener.scm
;;; note: source file /home/shegeley/Projects/guile-wayland/lisp/wayland/list.scm
;;;       newer than compiled /home/shegeley/.cache/guile/ccache/3.0-LE-8-4.6/home/shegeley/Projects/guile-wayland/lisp/wayland/list.scm.go
;;; compiling /home/shegeley/Projects/guile-wayland/lisp/wayland/list.scm
;;; compiling /home/shegeley/Projects/guile-wayland/lisp/wayland/util.scm
;;; WARNING: compilation of /home/shegeley/Projects/guile-wayland/lisp/wayland/util.scm failed:
;;; no code for module (wayland config)
;;; WARNING: compilation of /home/shegeley/Projects/guile-wayland/lisp/wayland/list.scm failed:
;;; no code for module (wayland config)
;;; WARNING: compilation of /home/shegeley/Projects/guile-wayland/lisp/wayland/listener.scm failed:
;;; no code for module (bytestructure-class)
;;; WARNING: compilation of /home/shegeley/Projects/guile-wayland/lisp/wayland/client.scm failed:
;;; In procedure dlopen: file "libguile-wayland.so", message "libguile-wayland.so: cannot open shared object file: No such file or directory"
;;; WARNING: compilation of /home/shegeley/Projects/guile-wayland/lisp/wayland.scm failed:
;;; no code for module (wayland config)
While executing meta-command:
no code for module (wayland config)
```
Also can't launch it interactively with `guix shell -D -f guix.scm -- guile`

```
shegeley@zone51 ~/Projects/guile-wayland$ guix shell -D -f guix.scm -- guile 
GNU Guile 3.0.9
Copyright (C) 1995-2023 Free Software Foundation, Inc.

Guile comes with ABSOLUTELY NO WARRANTY; for details type `,show w'.
This program is free software, and you are welcome to redistribute it
under certain conditions; type `,show c' for details.

Enter `,help' for help.
scheme@(guile-user)> ,m (wayland)
scheme@(wayland)> ,b
scheme@(wayland)> ,m (wayland client)
scheme@(wayland client)> ,b
scheme@(wayland client)> ,m (wayland display)
scheme@(wayland display)> ,b
scheme@(wayland display)> 
```

Comment (Z572) [reply repl.scm:2]: no `./pre-inst-env.in -- guile`，must invoke `guix shell -D -f guix.scm -- sh -c './bootstrap && ./configure && make'` , autotools will replace `@abs_top_srcdir@` to right value,and generate `pre-inst-env`, `pre-inst-env` will set env and run it args.

Comment (shegeley) [reply repl.scm:2]:
```
shegeley@zone51 ~/Projects/guile-wayland$ guix shell -D -f guix.scm -- sh -c './bootstrap && ./configure && make'
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force 
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, 'build-aux'.
libtoolize: copying file 'build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIRS, 'm4'.
libtoolize: copying file 'm4/libtool.m4'
libtoolize: copying file 'm4/ltoptions.m4'
libtoolize: copying file 'm4/ltsugar.m4'
libtoolize: copying file 'm4/ltversion.m4'
libtoolize: copying file 'm4/lt~obsolete.m4'
libtoolize: Consider adding '-I m4' to ACLOCAL_AMFLAGS in Makefile.am.
autoreconf: running: /gnu/store/kngd1h8b65kdnawd9mfxarwhkaxwq0lz-autoconf-2.69/bin/autoconf --force
autoreconf: running: /gnu/store/kngd1h8b65kdnawd9mfxarwhkaxwq0lz-autoconf-2.69/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
configure.ac:18: installing 'build-aux/compile'
configure.ac:8: installing 'build-aux/missing'
lisp/wayland/Makefile.am: installing 'build-aux/depcomp'
autoreconf: Leaving directory `.'
checking for a BSD-compatible install... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether UID '1002' is supported by ustar format... yes
checking whether GID '998' is supported by ustar format... yes
checking how to create a ustar tar archive... gnutar
checking whether make supports nested variables... (cached) yes
checking whether make supports the include directive... yes (GNU style)
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking dependency style of gcc... gcc3
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/grep
checking for egrep... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking dependency style of gcc... (cached) gcc3
checking for pkg-config... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
configure: checking for guile 3.0
configure: found guile 3.0
checking for guile-3.0... no
checking for guile3.0... no
checking for guile-3... no
checking for guile3... no
checking for guile... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/guile
checking for Guile version >= 3.0... 3.0.9
checking for guild... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/guild
checking for guile-config... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/guile-config
checking for Guile site directory... /gnu/store/4gvgcfdiz67wv04ihqfa8pqwzsb0qpv5-guile-3.0.9/share/guile/site/3.0
checking for Guile site-ccache directory using pkgconfig... /gnu/store/4gvgcfdiz67wv04ihqfa8pqwzsb0qpv5-guile-3.0.9/lib/guile/3.0/site-ccache
checking for Guile extensions directory... /gnu/store/4gvgcfdiz67wv04ihqfa8pqwzsb0qpv5-guile-3.0.9/lib/guile/3.0/extensions
checking for guile-3.0... yes
checking for wayland-server... yes
checking build system type... x86_64-pc-linux-gnu
checking host system type... x86_64-pc-linux-gnu
checking how to print strings... printf
checking for a sed that does not truncate output... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/sed
checking for fgrep... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/grep -F
checking for ld used by gcc... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/ld
checking if the linker (/gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/nm -B
checking the name lister (/gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking how to convert x86_64-pc-linux-gnu file names to x86_64-pc-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-pc-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/ld option to reload object files... -r
checking for file... file
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for a working dd... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/dd
checking how to truncate binary pipes... /gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/dd bs=4096 count=1
checking for mt... no
checking if : is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... yes
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/gnu/store/hwxvywi8nx18b4m3mwc8mpw3zgacz5xm-profile/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for shl_load... no
checking for shl_load in -ldld... no
checking for dlopen... yes
checking whether a program can dlopen itself... yes
checking whether a statically linked program can dlopen itself... no
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking wayland-server.h usability... yes
checking wayland-server.h presence... yes
checking for wayland-server.h... yes
checking libguile.h usability... no
checking libguile.h presence... no
checking for libguile.h... no
checking if (bytestructures guile) is available... yes
checking if (bytestructure-class) is available... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating lisp/wayland/Makefile
config.status: creating lisp/wayland/config.scm
config.status: creating build-aux/wayland-scanner-guile
config.status: creating pre-inst-env
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands
make  all-recursive
make[1]: Entering directory '/home/shegeley/Projects/guile-wayland'
Making all in lisp/wayland/
make[2]: Entering directory '/home/shegeley/Projects/guile-wayland/lisp/wayland'
make  all-am
make[3]: Entering directory '/home/shegeley/Projects/guile-wayland/lisp/wayland'
  GEN      config.go
wrote `config.go'
make[3]: Leaving directory '/home/shegeley/Projects/guile-wayland/lisp/wayland'
make[2]: Leaving directory '/home/shegeley/Projects/guile-wayland/lisp/wayland'
make[2]: Entering directory '/home/shegeley/Projects/guile-wayland'
make[2]: Leaving directory '/home/shegeley/Projects/guile-wayland'
make[1]: Leaving directory '/home/shegeley/Projects/guile-wayland'
shegeley@zone51 ~/Projects/guile-wayland$ 
```

This command won't allow to enter guix shell `[env]` at the very end. Just returns me to the bash.

Comment (Z572) [reply repl.scm:2]: `guix shell -D -f guix.scm -- sh -c './bootstrap && ./configure && make && ./pre-inst-env guile --listen'`

Comment (Z572) [reply repl.scm:2]: you can use this https://github.com/guile-wayland/guile-wayland/blob/ecadc0703333e42e35c55dcc343d882e05d2350a/repl.sh

Comment (Z572) [reply repl.scm:2]: oops, see https://github.com/guile-wayland/guile-wayland/blob/e098195c0ad4fbf17cf716a8cd9719ccde5d52d3/repl.sh

**Pull request closed without merging**